File: test/laz_writer_test.cpp
#include <cstring>
#include <string>

#include <catch2/catch.hpp>
#include <copc-lib/geometry/vector3.hpp>
#include <copc-lib/io/laz_reader.hpp>
#include <copc-lib/io/laz_writer.hpp>

using namespace copc;
using namespace std;

TEST_CASE("Laz Writer Config Tests", "[LAZ Writer]")
{
    GIVEN("A valid file_path")
    {
        SECTION("Default Config")
        {
            string file_path = "writer_test.laz";

            las::LazConfigWriter cfg(6);
            laz::LazFileWriter writer(file_path, cfg);

            // Test updating the values from the LasHeader getter
            writer.LazConfig()->LasHeader()->max = Vector3(10, 11, 12);

            auto las_header = writer.LazConfig()->LasHeader();
            REQUIRE(las_header->Scale().z == 0.01);
            REQUIRE(las_header->Offset().z == 0);
            REQUIRE(las_header->PointFormatId() == 6);
            REQUIRE(las_header->max == Vector3(10, 11, 12));
            REQUIRE(las_header->min == Vector3(0, 0, 0));

            writer.Close();

            REQUIRE_THROWS(las::LazConfigWriter(5));
            REQUIRE_THROWS(las::LazConfigWriter(9));
        }

        SECTION("Custom Config")
        {
            string file_path = "writer_test.laz";

            las::LazConfigWriter cfg(8, {2, 3, 4}, {-0.02, -0.03, -40.8});
            cfg.LasHeader()->file_source_id = 200;

            // Test checks on string attributes
            cfg.LasHeader()->SystemIdentifier("test_string");
            REQUIRE(cfg.LasHeader()->SystemIdentifier() == "test_string");
            REQUIRE_THROWS(cfg.LasHeader()->SystemIdentifier(std::string(33, 'a')));
            cfg.LasHeader()->GeneratingSoftware("test_string");
            REQUIRE(cfg.LasHeader()->GeneratingSoftware() == "test_string");
            REQUIRE_THROWS(cfg.LasHeader()->GeneratingSoftware(std::string(33, 'a')));

            laz::LazFileWriter writer(file_path, cfg);

            auto las_header = writer.LazConfig()->LasHeader();
            REQUIRE(las_header->file_source_id == 200);
            REQUIRE(las_header->PointFormatId() == 8);
            REQUIRE(las_header->Scale().x == 2);
            REQUIRE(las_header->Offset().x == -0.02);
            REQUIRE(las_header->Scale().y == 3);
            REQUIRE(las_header->Offset().y == -0.03);
            REQUIRE(las_header->Scale().z == 4);
            REQUIRE(las_header->Offset().z == -40.8);

            writer.Close();
        }

        SECTION("WKT")
        {
            string file_path = "writer_test.laz";

            las::LazConfigWriter cfg(6, Vector3::DefaultScale(), Vector3::DefaultOffset(), "TEST_WKT");
            laz::LazFileWriter writer(file_path, cfg);

            REQUIRE(writer.LazConfig()->Wkt() == "TEST_WKT");

            writer.Close();

            laz::LazFileReader reader(file_path);
            REQUIRE(reader.LazConfig().Wkt() == "TEST_WKT");
        }
    }

    GIVEN("An invalid filepath")
    {
        REQUIRE_THROWS(laz::LazFileWriter("invalid_path/writer_test.laz", las::LazConfigWriter(6)));
    }

    GIVEN("A valid output stream")
    {
        SECTION("Default Config")
        {
            stringstream out_stream;

            las::LazConfigWriter cfg(6);
            laz::LazWriter writer(out_stream, cfg);

            auto las_header = writer.LazConfig()->LasHeader();
            REQUIRE(las_header->Scale().z == 0.01);
            REQUIRE(las_header->Offset().z == 0);
            REQUIRE(las_header->PointFormatId() == 6);

            writer.Close();

            lazperf::reader::generic_file f(out_stream);
            REQUIRE(f.pointCount() == 0);
            REQUIRE(f.header().scale.z == 0.01);
            REQUIRE(f.header().offset.z == 0);
            REQUIRE(f.header().point_format_id == 6);
        }

        SECTION("Custom Config")
        {
            stringstream out_stream;

            las::LazConfigWriter cfg(8, {2, 3, 4}, {-0.02, -0.03, -40.8});
            cfg.LasHeader()->file_source_id = 200;
            laz::LazWriter writer(out_stream, cfg);

            auto las_header = writer.LazConfig()->LasHeader();
            REQUIRE(las_header->file_source_id == 200);
            REQUIRE(las_header->PointFormatId() == 8);
            REQUIRE(las_header->Scale().x == 2);
            REQUIRE(las_header->Offset().x == -0.02);
            REQUIRE(las_header->Scale().y == 3);
            REQUIRE(las_header->Offset().y == -0.03);
            REQUIRE(las_header->Scale().z == 4);
            REQUIRE(las_header->Offset().z == -40.8);

            writer.Close();

            lazperf::reader::generic_file f(out_stream);
            REQUIRE(f.pointCount() == 0);
            REQUIRE(f.header().file_source_id == 200);
            REQUIRE(f.header().point_format_id == 8);
            REQUIRE(f.header().scale.x == 2);
            REQUIRE(f.header().offset.x == -0.02);
            REQUIRE(f.header().scale.y == 3);
            REQUIRE(f.header().offset.y == -0.03);
            REQUIRE(f.header().scale.z == 4);
            REQUIRE(f.header().offset.z == -40.8);
        }

        SECTION("WKT")
        {
            stringstream out_stream;

            las::LazConfigWriter cfg(6, {}, {}, "TEST_WKT");
            laz::LazWriter writer(out_stream, cfg);

            REQUIRE(writer.LazConfig()->Wkt() == "TEST_WKT");

            writer.Close();

            laz::LazReader reader(&out_stream);
            REQUIRE(reader.LazConfig().Wkt() == "TEST_WKT");
        }
    }
}

TEST_CASE("LAZ Writer EBs", "[LAZ Writer]")
{
    SECTION("Data type 0")
    {
        stringstream out_stream;
        las::EbVlr eb_vlr(1); // Always initialize with the ebCount constructor
        // don't make ebfields yourself unless you set their names correctly
        eb_vlr.items[0].data_type = 0;
        eb_vlr.items[0].options = 4;
        las::LazConfigWriter cfg(7, {}, {}, {}, eb_vlr);
        laz::LazWriter writer(out_stream, cfg);

        REQUIRE(writer.LazConfig()->LasHeader()->PointRecordLength() == 40); // 36 + 4

        writer.Close();

        laz::LazReader reader(&out_stream);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items.size() == 1);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items[0].data_type == 0);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items[0].options == 4);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items[0].name == "FIELD_0");
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items[0].maxval[2] == 0);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items[0].minval[2] == 0);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items[0].offset[2] == 0);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items[0].scale[2] == 0);
        REQUIRE(reader.LazConfig().LasHeader().PointRecordLength() == 40);
    }

    SECTION("Data type 29")
    {
        stringstream out_stream;
        las::EbVlr eb_vlr(1);
        eb_vlr.items[0].data_type = 29;
        las::LazConfigWriter cfg(7, {}, {}, {}, eb_vlr);
        laz::LazWriter writer(out_stream, cfg);

        REQUIRE(writer.LazConfig()->LasHeader()->PointRecordLength() == 48); // 36 + 12

        writer.Close();

        laz::LazReader reader(&out_stream);
        REQUIRE(reader.LazConfig().ExtraBytesVlr().items.size() == 1);
        REQUIRE(reader.LazConfig().LasHeader().PointRecordLength() == 48);
    }
}

TEST_CASE("LAZ Write Points", "[LAZ Writer]")
{
    string file_path = "writer_test.laz";

    las::LazConfigWriter cfg(6);
    laz::LazFileWriter writer(file_path, cfg);

    // Create a Points object based on the LAS header
    las::Points points(*cfg.LasHeader());
    // Populate the points
    // Create a point with a given point format
    // The use of las::Point constructor is strongly discouraged, instead use las::Points::CreatePoint
    auto point = points.CreatePoint();
    // point has getters/setters for all attributes
    point->X(0);
    point->Y(0);
    point->Z(0);
    points.AddPoint(point);
    point = points.CreatePoint();
    // point has getters/setters for all attributes
    point->X(11);
    point->Y(12);
    point->Z(13);
    points.AddPoint(point);

    writer.WritePoints(points);
    writer.WritePoints(points);
    writer.Close();

    // Validate
    laz::LazFileReader reader(file_path);
    auto read_points = reader.GetPoints();
    REQUIRE(read_points.Size() == 4);
    REQUIRE(read_points.Get(0)->X() == 0);
    REQUIRE(read_points.Get(0)->Y() == 0);
    REQUIRE(read_points.Get(0)->Z() == 0);
    REQUIRE(read_points.Get(1)->X() == 11);
    REQUIRE(read_points.Get(1)->Y() == 12);
    REQUIRE(read_points.Get(1)->Z() == 13);
    REQUIRE(read_points.Get(2)->X() == 0);
    REQUIRE(read_points.Get(2)->Y() == 0);
    REQUIRE(read_points.Get(2)->Z() == 0);
    REQUIRE(read_points.Get(3)->X() == 11);
    REQUIRE(read_points.Get(3)->Y() == 12);
    REQUIRE(read_points.Get(3)->Z() == 13);
}

File: test/laz_writer_test.py
import copclib as copc
import pytest
import os

from .utils import get_autzen_file, get_data_dir


def test_writer_config():
    # Given a valid file path
    file_path = os.path.join(get_data_dir(), "writer_test.laz")

    # Default config
    cfg = copc.LazConfigWriter(6)
    writer = copc.LazWriter(file_path, cfg)

    las_header = writer.laz_config.las_header
    assert las_header.scale == [0.01, 0.01, 0.01]
    assert las_header.offset == [0.0, 0.0, 0.0]
    assert las_header.point_format_id == 6

    str(cfg)

    with pytest.raises(RuntimeError):
        assert copc.LazConfigWriter(5)
        assert copc.LazConfigWriter(9)

    writer.Close()

    # Given an invalid file path
    with pytest.raises(RuntimeError):
        assert copc.LazWriter(
            os.path.join(get_data_dir(), "invalid_path", "non_existant_file.copc.laz"),
            cfg,
        )

    # Custom config

    cfg = copc.LazConfigWriter(8, [2, 3, 4], [-0.02, -0.03, -40.8])
    cfg.las_header.file_source_id = 200

    # Test LasHeaderConfig attributes
    cfg.las_header.creation_day = 18
    assert cfg.las_header.creation_day == 18
    cfg.las_header.creation_year = 11
    assert cfg.las_header.creation_year == 11

    # Test checks on string attributes
    cfg.las_header.guid = "test_string"
    assert cfg.las_header.guid == "test_string"
    with pytest.raises(RuntimeError):
        cfg.las_header.guid = "a" * 17

    cfg.las_header.system_identifier = "test_string"
    assert cfg.las_header.system_identifier == "test_string"
    with pytest.raises(RuntimeError):
        cfg.las_header.system_identifier = "a" * 33

    cfg.las_header.generating_software = "test_string"
    assert cfg.las_header.generating_software == "test_string"
    with pytest.raises(RuntimeError):
        cfg.las_header.generating_software = "a" * 33

    writer = copc.LazWriter(file_path, cfg)

    las_header = writer.laz_config.las_header
    assert las_header.file_source_id == 200
    assert las_header.point_format_id == 8
    assert las_header.scale == [2.0, 3.0, 4.0]
    assert las_header.offset == [-0.02, -0.03, -40.8]

    writer.Close()

    # WKT
    cfg = copc.LazConfigWriter(6, wkt="TEST_WKT")
    writer = copc.LazWriter(file_path, cfg)

    assert writer.laz_config.wkt == "TEST_WKT"

    writer.Close()

    reader = copc.LazReader(file_path)
    assert reader.laz_config.wkt == "TEST_WKT"

    # Copy
    orig = copc.LazReader(get_autzen_file())

    cfg = orig.laz_config
    writer = copc.LazWriter(file_path, cfg)

    writer.laz_config.las_header.min = (1, 1, 1)
    writer.laz_config.las_header.max = (50, 50, 50)

    assert writer.laz_config.las_header.min == (1, 1, 1)
    assert writer.laz_config.las_header.max == (50, 50, 50)

    writer.Close()

    reader = copc.LazReader(file_path)
    assert (
        reader.laz_config.las_header.file_source_id
        == orig.laz_config.las_header.file_source_id
    )
    assert (
        reader.laz_config.las_header.global_encoding
        == orig.laz_config.las_header.global_encoding
    )
    assert (
        reader.laz_config.las_header.creation_day
        == orig.laz_config.las_header.creation_day
    )
    assert (
        reader.laz_config.las_header.creation_year
        == orig.laz_config.las_header.creation_year
    )
    assert (
        reader.laz_config.las_header.file_source_id
        == orig.laz_config.las_header.file_source_id
    )
    assert (
        reader.laz_config.las_header.point_format_id
        == orig.laz_config.las_header.point_format_id
    )
    assert (
        reader.laz_config.las_header.point_record_length
        == orig.laz_config.las_header.point_record_length
    )
    assert reader.laz_config.las_header.point_count == 0
    assert reader.laz_config.las_header.min == (1, 1, 1)
    assert reader.laz_config.las_header.max == (50, 50, 50)

    # Update
    min1 = (-800, 300, 800)
    max1 = (5000, 8444, 3333)
    min2 = (-20, -30, -40)
    max2 = (20, 30, 40)
    points_by_return = list(range(15))

    cfg = copc.LazConfigWriter(6, wkt="TEST_WKT")
    cfg.las_header.min = min1
    cfg.las_header.max = max1
    cfg.copc_info.spacing = 10
    writer = copc.LazWriter(file_path, cfg)

    assert writer.laz_config.las_header.min == min1
    assert writer.laz_config.las_header.max == max1
    assert writer.laz_config.las_header.points_by_return == [0] * 15

    with pytest.raises(TypeError):
        writer.laz_config.las_header.points_by_return = [20] * 800

    writer.laz_config.las_header.min = min2
    writer.laz_config.las_header.max = max2
    writer.laz_config.las_header.points_by_return = points_by_return

    assert writer.laz_config.las_header.min == min2
    assert writer.laz_config.las_header.max == max2
    assert writer.laz_config.las_header.points_by_return == points_by_return

    writer.Close()

    reader = copc.LazReader(file_path)
    assert reader.laz_config.las_header.min == min2
    assert reader.laz_config.las_header.max == max2
    assert reader.laz_config.las_header.points_by_return == points_by_return


def test_writer_copy():
    # Given a valid file path
    file_path = os.path.join(get_data_dir(), "writer_test.copc.laz")

    reader = copc.FileReader(get_autzen_file())

    cfg = reader.copc_config
    writer = copc.LazWriter(file_path, cfg)

    num_pts = 0
    for node in reader.GetNodesAtResolution(10):
        # only write/compare compressed data or otherwise tests take too long
        writer.WritePointsCompressed(
            reader.GetPointDataCompressed(node), node.point_count
        )
        num_pts += node.point_count

    writer.Close()

    # validate
    new_reader = copc.LazReader(file_path)
    points = new_reader.GetPoints()
    assert len(points) == num_pts


def test_write():
    file_path = os.path.join(get_data_dir(), "writer_test.copc.laz")
    cfg = copc.LazConfigWriter(6)
    writer = copc.LazWriter(file_path, cfg)

    points = copc.Points(cfg.las_header)
    p = points.CreatePoint()
    p.x = 0
    p.y = 0
    p.z = 0
    points.AddPoint(p)
    p = points.CreatePoint()
    p.x = 11
    p.y = 12
    p.z = 13
    points.AddPoint(p)

    writer.WritePoints(points)
    writer.WritePoints(points)
    writer.Close()

    reader = copc.LazReader(file_path)
    read_points = reader.GetPoints()
    assert len(read_points) == 4
    assert read_points.x == [0, 11, 0, 11]
    assert read_points.y == [0, 12, 0, 12]
    assert read_points.z == [0, 13, 0, 13]
